src/ControlSystem/TimescaleTuner.cpp

```diff
@@ -0,0 +1,118 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ControlSystem/TimescaleTuner.hpp"
+
+TimescaleTuner::TimescaleTuner(DataVector initial_timescale,
+                               double max_timescale, double min_timescale,
+                               double max_threshold, double min_threshold,
+                               double increase_factor,
+                               double decrease_factor) noexcept
+    : timescale_{initial_timescale},
+      max_threshold_{max_threshold},
+      min_threshold_{min_threshold},
+      max_timescale_(max_timescale),
+      min_timescale_{min_timescale},
+      increase_factor_{increase_factor},
+      decrease_factor_{decrease_factor} {
+  for (auto t_scale : timescale_) {
+    if (t_scale <= 0.0) {
+      ERROR("Initial timescale must be > 0");
+    }
+  }
+
+  if (decrease_factor_ > 1.0 or decrease_factor <= 0.0) {
+    ERROR("The specified decrease_factor "
+          << decrease_factor_ << " must satisfy 0 < decrease_factor <= 1");
+  }
+  if (increase_factor_ < 1.0 or increase_factor_ < decrease_factor_) {
+    ERROR("The specified increase factor "
+          << increase_factor_ << " must satisfy decrease_factor("
+          << decrease_factor_ << ") < 1.0 <= increase_factor");
+  }
+  if (min_timescale_ <= 0.0) {
+    ERROR("The specified minimum timescale " << min_timescale_
+                                             << " must be > 0");
+  }
+  if (max_timescale_ <= min_timescale_) {
+    ERROR("The maximum timescale "
+          << max_timescale_
+          << " must be > than the specified minimum timescale "
+          << min_timescale_);
+  }
+  if (min_threshold_ <= 0.0) {
+    ERROR("The specified minimum timescale " << min_timescale_
+                                             << " must be > 0");
+  }
+  if (max_threshold_ <= min_threshold_) {
+    ERROR("The maximum threshold "
+          << max_threshold_
+          << " must be > than the specified minimum threshold "
+          << min_threshold_);
+  }
+}
+
+void TimescaleTuner::suggested_timescale_if_in_allowable_range(
```

Comment:
Pleas test this for `timescale < min_timescale_`, `min_timescale_ < timescale < max_timescale_`, and `max_timescale_ < timescale`.